### DISCUSSION TITLE
It is suggested to change the CDN of Swagger and Redoc

### DISCUSSION THREAD

**Discussion by haitanghuadeng** [fastapi/fastapi#8500] [Questions] [answered]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
https://cdn.jsdelivr.net/npm/swagger-ui-dist@4/swagger-ui.css
https://cdn.jsdelivr.net/npm/swagger-ui-dist@4/swagger-ui-bundle.js
```

### Description

Two years later, I still have trouble loading 'Swagger' resources. First step, I looked up the article about 'FastAPI' and 'Swagger' not loading. Some friends suggest using 'CSS' and 'JavaScript' files stored in the 'static' folder of your project, And modify the Python 'Lib/site-package/fastapi/openapi/docs.py' the 'url' for the relative path. However, I don't think it's a good idea to modify your own source code.

### Operating System

Windows

### Operating System Details

Windows11(22H2.25120.1010)

### FastAPI Version

0.78.0

### Python Version

3.10.0

### Additional Context

So, when I looked through the official use case of 'Swagger', I realized that it might have something to do with 'CDN'. Therefore, it is recommended to use a modified CDN. Thank you very much!
CSS Link https://unpkg.com/swagger-ui-dist@4/swagger-ui.css
JavaScript Link https://unpkg.com/swagger-ui-dist@4/swagger-ui-bundle.js

**Comment by coolmian** [comment]
**Here's a quick solution**

You can read [#4307]( https://github.com/tiangolo/fastapi/issues/4307)

Download from:
https://cdnjs.cloudflare.com/ajax/libs/swagger-ui/3.52.4/swagger-ui-bundle.js
https://cdnjs.cloudflare.com/ajax/libs/swagger-ui/3.52.4/swagger-ui.css

Create dir _./static/swagger-ui/_ on your project, and put these files in.


Provides static file services：
```python
from fastapi import applications
from fastapi.staticfiles import StaticFiles
from fastapi.openapi.docs import get_swagger_ui_html

app.mount("/static", StaticFiles(directory="./static"), name="static")
def swagger_monkey_patch(*args, **kwargs):
    return get_swagger_ui_html(
        *args,
        **kwargs,
        swagger_js_url="/static/swagger-ui/swagger-ui-bundle.js",
        swagger_css_url="/static/swagger-ui/swagger-ui.css",
    )
applications.get_swagger_ui_html = swagger_monkey_patch
```

_Originally posted by @SangYuming in https://github.com/tiangolo/fastapi/issues/4307#issuecomment-997774036_

**Comment by ahmetoner** [comment]
I created a repository which is called 'Fastapi Offline Swagger UI' to which is only download js and css files for offline usage. You can find here [ahmetoner/fastapi-offline-swagger-ui](https://github.com/ahmetoner/fastapi-offline-swagger-ui) 


### Install
You can install this software via pip or poetry.

```sh
pip3 install git+https://github.com/ahmetoner/fastapi-offline-swagger-ui
```

or pyproject.toml

```sh

[tool.poetry.dependencies]
...
fastapi-offline-swagger-ui = {git = "https://github.com/ahmetoner/fastapi-offline-swagger-ui"}

```


### Usage Examples
After installing the module, create a python file. Copy and run following.
```py
from fastapi import FastAPI, File, UploadFile, Query, applications
from fastapi.responses import StreamingResponse, RedirectResponse
from fastapi.staticfiles import StaticFiles
from fastapi.openapi.docs import get_swagger_ui_html
import fastapi_offline_swagger_ui

app = FastAPI()

''' This following code block necessary to switch offline cdn files via fastapi_offline_swagger_ui module
'''
assets_path = fastapi_offline_swagger_ui.__path__[0]
if path.exists(assets_path + "/swagger-ui.css") and path.exists(assets_path + "/swagger-ui-bundle.js"):
    app.mount("/assets", StaticFiles(directory=assets_path), name="static")
    def swagger_monkey_patch(*args, **kwargs):
        return get_swagger_ui_html(
            *args,
            **kwargs,
            swagger_favicon_url="",
            swagger_css_url="/assets/swagger-ui.css",
            swagger_js_url="/assets/swagger-ui-bundle.js",
        )
    applications.get_swagger_ui_html = swagger_monkey_patch

```

### Live Example
Alternatively you can browse [ahmetoner/whisper-asr-webservice](https://github.com/ahmetoner/whisper-asr-webservice/) repository to see how it is implemented.

**Comment by gaby** [comment]
@tiangolo Any reason this is not built-into `FastAPI` ? This project is used in a lot of offline/homelab environments. A lot of open source tools also use FastAPI and when deployed on home-labs the API's don't work at all.

**Comment by YuriiMotov** [comment, marked as answer]
Official solution: https://fastapi.tiangolo.com/how-to/custom-docs-ui-assets/
